metrics-core/src/main/java/com/codahale/metrics/LockFreeExponentiallyDecayingReservoir.java

```diff
@@ -0,0 +1,247 @@
+ * however it provides looser guarantees while completely avoiding locks.
+ *
+ * Looser guarantees:
+ * <ul>
+ *     <li> Updates which occur concurrently with rescaling may be discarded if the orphaned state node is updated after
+ *     rescale has replaced it. This condition has a greater probability as the rescale interval is reduced due to the
+ *     increased frequency of rescaling. {@link #rescaleThresholdNanos} values below 30 seconds are not recommended.
+ *     <li> Given a small rescale threshold, updates may attempt to rescale into a new bucket, but lose the CAS race
+ *     and update into a newer bucket than expected. In these cases the measurement weight is reduced accordingly.
+ *     <li>In the worst case, all concurrent threads updating the reservoir may attempt to rescale rather than
+ *     a single thread holding an exclusive write lock. It's expected that the configuration is set such that
+ *     rescaling is substantially less common than updating at peak load. Even so, when size is reasonably small
+ *     it can be more efficient to rescale than to park and context switch.
+ * </ul>
+ *
+ * @author <a href="mailto:ckozak@ckozak.net">Carter Kozak</a>
+ */
+public final class LockFreeExponentiallyDecayingReservoir implements Reservoir {
+
+    private static final double SECONDS_PER_NANO = .000000001D;
+    private static final AtomicReferenceFieldUpdater<LockFreeExponentiallyDecayingReservoir, State> stateUpdater =
+            AtomicReferenceFieldUpdater.newUpdater(LockFreeExponentiallyDecayingReservoir.class, State.class, "state");
+
+    private final double alphaNanos;
+    private final int size;
+    private final long rescaleThresholdNanos;
+    private final Clock clock;
+
+    private volatile State state;
+
+    private final class State {
+        private final long startTick;
+        // Count is updated after samples are successfully added to the map.
+        private final AtomicLong count;
+        private final ConcurrentSkipListMap<Double, WeightedSample> values;
+
+        State(long startTick, AtomicLong count, ConcurrentSkipListMap<Double, WeightedSample> values) {
+            this.startTick = startTick;
+            this.values = values;
+            this.count = count;
+        }
+
+        private void update(long value, long timestampNanos) {
+            double itemWeight = weight(timestampNanos - startTick);
+            double priority = itemWeight / ThreadLocalRandom.current().nextDouble();
+            long currentCount = count.get();
+            if (currentCount < size || values.firstKey() < priority) {
+                addSample(priority, value, itemWeight);
+            }
+        }
+
+        private void addSample(double priority, long value, double itemWeight) {
+            if (values.putIfAbsent(priority, new WeightedSample(value, itemWeight)) == null
+                    && count.incrementAndGet() > size) {
```

Comment:
Might be nice to avoid doing the increment if the count is already >= size? The only reason to increment it is to detect when it's large enough, so if the previous check: `currentCount < size` fails, there's no need to increment it further.

You could just extract that check into a variable and pass it in to the function

Reply:
In testing this didn't produce a measurable difference, however I think it could allow us to switch `count` from a 64-bit value to a 32-bit value.

What do you think? https://github.com/carterkozak/metrics/pull/3

Reply:
LGTM!